src/analysis/processing/qgsalgorithmextractlabels.cpp

```diff
@@ -0,0 +1,372 @@
+          labelAlignment = QStringLiteral( "center" );
+          break;
+
+        case QgsPalLayerSettings::QgsPalLayerSettings::MultiLeft:
+          labelAlignment = QStringLiteral( "left" );
+          break;
+
+        case QgsPalLayerSettings::QgsPalLayerSettings::MultiJustify:
+          labelAlignment = QStringLiteral( "justify" );
+          break;
+
+        case QgsPalLayerSettings::MultiFollowPlacement:
+          switch ( label->getQuadrant() )
+          {
+            case pal::LabelPosition::QuadrantAboveLeft:
+            case pal::LabelPosition::QuadrantLeft:
+            case pal::LabelPosition::QuadrantBelowLeft:
+              labelAlignment = QStringLiteral( "right" );
+              break;
+
+            case pal::LabelPosition::QuadrantAbove:
+            case pal::LabelPosition::QuadrantOver:
+            case pal::LabelPosition::QuadrantBelow:
+              labelAlignment = QStringLiteral( "center" );
+              break;
+
+            case pal::LabelPosition::QuadrantAboveRight:
+            case pal::LabelPosition::QuadrantRight:
+            case pal::LabelPosition::QuadrantBelowRight:
+              labelAlignment = QStringLiteral( "left" );
+              break;
+          }
+          break;
+      }
+
+      const double labelRotation = label->getAlpha() * 180 / M_PI;
+
+      const QFont font = labelFeature->definedFont();
+      const QString fontFamily = font.family();
+      const QString fontStyle = font.styleName();
+      const double fontSize = static_cast<double>( font.pixelSize() ) * 72 / context.painter()->device()->logicalDpiX();
+      const bool fontItalic = font.italic();
+      const bool fontBold = font.bold();
+      const double fontLetterSpacing = font.letterSpacing();
+      const double fontWordSpacing = font.wordSpacing();
+
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::MultiLineHeight ) )
+      {
+        QgsTextFormat format = labelSettings.format();
+        format.setLineHeight( dataDefinedValues.value( QgsPalLayerSettings::MultiLineHeight ).toDouble() );
+        labelSettings.setFormat( format );
+      }
+
+      QgsTextFormat format = labelSettings.format();
+      double formatLineHeight = format.lineHeight();
+
+      QgsAttributes attributes;
+      attributes << mMapLayerNames->value( layerId ) << fid << labelText << labelAlignment << labelRotation
+                 << fontFamily << fontStyle << fontSize << fontItalic << fontBold << fontLetterSpacing << fontWordSpacing
+                 << formatLineHeight;
+
+      double x = label->getX();
+      double y = label->getY();
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::PositionX ) )
+      {
+        x = dataDefinedValues.value( QgsPalLayerSettings::PositionX ).toDouble();
+      }
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::PositionY ) )
+      {
+        y = dataDefinedValues.value( QgsPalLayerSettings::PositionY ).toDouble();
+      }
```

Comment:
this shouldn't be required -- pal has already taken it into account

Reply:
@nyalldawson , in my tests, I saw that label->getX() and  label->getY() would return the non manually set position, hence why this was needed.

Reply:
I'm almost 100% it's not required -- otherwise we'd see similar logic in place for rendering labels, and things like QgsLabelSearchTree::insertLabel would be broken for data-defined placements